File: mobile/app/src/main/java/com/cryptopulse/app/domain/utils/SymbolResolver.kt
package com.cryptopulse.app.domain.utils

import java.util.Locale

data class ResolvedSymbol(
    val symbol: String,
    val baseAsset: String,
    val quoteAsset: String
)

object SymbolResolver {
    private val KNOWN_QUOTES = listOf("USDT", "BUSD", "USD", "USDC", "BTC", "ETH")

    fun resolve(inputSymbol: String, defaultQuote: String = "USDT"): ResolvedSymbol {
        require(inputSymbol.isNotBlank()) { "Invalid symbol input: must be a non-empty string." }

        val cleaned = inputSymbol.trim().uppercase(Locale.ROOT).replace(Regex("[/\\s\\-_]"), "")

        for (quote in KNOWN_QUOTES) {
            if (cleaned.endsWith(quote) && cleaned.length > quote.length) {
                val base = cleaned.substring(0, cleaned.length - quote.length)
                return ResolvedSymbol(
                    symbol = cleaned,
                    baseAsset = base,
                    quoteAsset = quote
                )
            }
        }

        val quote = defaultQuote.uppercase(Locale.ROOT)
        return ResolvedSymbol(
            symbol = "$cleaned$quote",
            baseAsset = cleaned,
            quoteAsset = quote
        )
    }

    fun toCacheKey(inputSymbol: String, defaultQuote: String = "USDT"): String {
        return resolve(inputSymbol, defaultQuote).symbol
    }
}

File: backend/src/utils/SymbolResolver.ts
export interface ResolvedSymbol {
  symbol: string;     // e.g. "BTCUSDT"
  baseAsset: string;  // e.g. "BTC"
  quoteAsset: string; // e.g. "USDT"
}

export class SymbolResolver {
  private static KNOWN_QUOTES = ["USDT", "BUSD", "USD", "USDC", "BTC", "ETH"];

  /**
   * Normalizes any input symbol string (e.g. "btc", "BTC/USDT", "BTC-USDT", "BTCUSDT")
   * into a standardized upper-case symbol format and extracts base/quote assets.
   */
  public static resolve(inputSymbol: string, defaultQuote: string = "USDT"): ResolvedSymbol {
    if (!inputSymbol || typeof inputSymbol !== "string") {
      throw new Error("Invalid symbol input: must be a non-empty string.");
    }

    const cleaned = inputSymbol.trim().toUpperCase().replace(/[/\s_-]/g, "");

    // If cleaned string ends with any known quote currency
    for (const quote of this.KNOWN_QUOTES) {
      if (cleaned.endsWith(quote) && cleaned.length > quote.length) {
        const base = cleaned.slice(0, cleaned.length - quote.length);
        return {
          symbol: cleaned,
          baseAsset: base,
          quoteAsset: quote,
        };
      }
    }

    // Default case: Treat input as base asset (e.g. "BTC" -> "BTCUSDT")
    const quote = defaultQuote.toUpperCase();
    return {
      symbol: `${cleaned}${quote}`,
      baseAsset: cleaned,
      quoteAsset: quote,
    };
  }

  /**
   * Returns standard cache key representation for exchange metadata maps.
   */
  public static toCacheKey(inputSymbol: string, defaultQuote: string = "USDT"): string {
    return this.resolve(inputSymbol, defaultQuote).symbol;
  }
}
